domain/keymanager/state/state.go

```diff
@@ -280,48 +529,133 @@ WHERE user_id = $userId.user_id
-// public key data or comment. If no user exists for the given id a nill error
-// will be returned.
+// list of public keys on the given model. keyIds represent one of the keys
+// fingerprint, public key data or comment.
+// The following errors can be expected:
+// - [accesserrors.UserNotFound] - When the user does not exist.
+// - [modelerrors.NotFound] - When the model does not exist.
 func (s *State) DeletePublicKeysForUser(
 	ctx context.Context,
-	id user.UUID,
+	modelUUID model.UUID,
+	userUUID user.UUID,
 	keyIds []string,
 ) error {
 	db, err := s.DB()
 	if err != nil {
-		return errors.Trace(err)
+		return err
 	}
 
-	stmt, err := s.Prepare(`
-DELETE FROM user_public_ssh_key
-WHERE user_id = $userId.user_id
+	userUUIDVal := userUUIDValue{UUID: userUUID.String()}
+	modelUUIDVal := modelUUIDValue{UUID: modelUUID.String()}
+
+	input := make(sqlair.S, 0, len(keyIds))
+	for _, keyId := range keyIds {
+		input = append(input, keyId)
+	}
+
+	findKeysStmt, err := s.Prepare(`
+SELECT (id) AS (&userPublicKeyId.*)
+FROM user_public_ssh_key
+WHERE user_uuid = $userUUIDValue.user_uuid
 AND (comment IN ($S[:])
   OR fingerprint IN ($S[:])
   OR public_key IN ($S[:]))
-`, sqlair.S{}, userId{})
+`, userUUIDVal, userPublicKeyId{}, input)
+	if err != nil {
+		return errors.Errorf(
+			"preparing find keys statement when deleting public keys for user %q on model %q: %w",
+			userUUID, modelUUID, err,
+		)
+	}
 
+	deleteFromModelStmt, err := s.Prepare(`
+DELETE FROM model_authorized_keys
+WHERE user_public_ssh_key_id IN ($userPublicKeyIds[:])
+AND model_uuid = $modelUUIDValue.model_uuid
+`, modelUUIDVal, userPublicKeyIds{})
 	if err != nil {
-		return fmt.Errorf(
-			"preparing delete statement for removing public keys for user %q: %w",
-			id, err,
+		return errors.Errorf(
+			"preparing delete keys statement when deleting public keys for user %q on model %q: %w",
+			userUUID, modelUUID, err,
 		)
 	}
 
-	input := make(sqlair.S, 0, len(keyIds))
-	for _, keyId := range keyIds {
-		input = append(input, keyId)
+	// deleteUnusedUserKeys is here to clean up any public keys for a user that
+	// are not being referenced by a model.
+	deleteUnusedUserKeys, err := s.Prepare(`
+DELETE FROM user_public_ssh_key
+WHERE user_uuid = $userUUIDValue.user_uuid
+AND id IN (SELECT id
+           FROM user_public_ssh_key AS upsk
+           LEFT JOIN model_authorized_keys AS mak ON upsk.id = mak.user_public_ssh_key_id
+           GROUP BY (id)
+           HAVING count(user_public_ssh_key_id) == 0)
+`, userUUIDVal)
+
+	if err != nil {
+		return errors.Errorf(
+			"preparing delete unused user keys statement when deleting public keys for user %q on model %q: %w",
+			userUUID, modelUUID, err,
+		)
 	}
 
-	userId := userId{id.String()}
+	foundKeyIds := userPublicKeyIds{}
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		return tx.Query(ctx, stmt, userId, input).Run()
+		err := s.checkUserExists(ctx, userUUID, tx)
```

Comment:
For deletion, not sure if we have to check the existence of users and models.

Reply:
We should always check our pre-conditions and fail! This is a big part of transaction retrying. Granted in this case it would all work regardless of if the user existed. But the call sites that a fundamentally ending up here need this validation. If we had an API that didn't return not found for a user when that is the subject of your operation would be confusing for both the caller/user and also potentially introduce security problems into the system.

TLDR Always check and validate the subject preconditions for any call.